BlueDBOnDisk/src/main/java/org/bluedb/disk/file/BlueObjectOutput.java

```diff
@@ -16,55 +19,94 @@
+				metadata.put(BlueFileMetadataKey.ENCRYPTION_VERSION_KEY, encryptionService.getCurrentEncryptionVersionKey());
+			}
+		} catch (Throwable t) {
 			close();
 			throw new BlueDbException(t.getMessage(), t);
 		}
 	}
 
-	protected static <T> BlueObjectOutput<T> getTestOutput(Path path, BlueSerializer serializer, DataOutputStream dataOutputStream) {
-		return new BlueObjectOutput<T>(path, serializer, dataOutputStream);
+	protected static <T> BlueObjectOutput<T> getTestOutput(Path path, BlueSerializer serializer, EncryptionServiceWrapper encryptionService, DataOutputStream dataOutputStream) {
+		return new BlueObjectOutput<>(path, serializer, encryptionService, dataOutputStream);
 	}
 
-	private BlueObjectOutput(Path path, BlueSerializer serializer, DataOutputStream dataOutputStream) {
-		LockManager<Path> lockManager = new LockManager<Path>();
+	private BlueObjectOutput(Path path, BlueSerializer serializer, EncryptionServiceWrapper encryptionService, DataOutputStream dataOutputStream) {
+		LockManager<Path> lockManager = new LockManager<>();
 		lock = lockManager.acquireWriteLock(path);
 		this.path = path;
 		this.serializer = serializer;
+		this.encryptionService = encryptionService;
 		this.dataOutputStream = dataOutputStream;
+
+		metadata = new BlueFileMetadata();
+		if (encryptionService.isEncryptionEnabled()) {
+			metadata.put(BlueFileMetadataKey.ENCRYPTION_VERSION_KEY, encryptionService.getCurrentEncryptionVersionKey());
+		}
 	}
-	
+
 	public static <T> BlueObjectOutput<T> createWithoutLockOrSerializer(Path path) throws BlueDbException {
-		return createWithoutLock(path, null);
+		return createWithoutLock(path, null, new EncryptionServiceWrapper(null));
 	}
-	
-	public static <T> BlueObjectOutput<T> createWithoutLock(Path path, BlueSerializer serializer) throws BlueDbException {
-		return new BlueObjectOutput<>(path, serializer);
+
+	public static <T> BlueObjectOutput<T> createWithoutLock(Path path, BlueSerializer serializer, EncryptionServiceWrapper encryptionService) throws BlueDbException {
+		return new BlueObjectOutput<>(path, serializer, encryptionService);
 	}
 
-	private BlueObjectOutput(Path path, BlueSerializer serializer) throws BlueDbException {
+	private BlueObjectOutput(Path path, BlueSerializer serializer, EncryptionServiceWrapper encryptionService) throws BlueDbException {
 		try {
 			this.lock = null;
 			this.path = path;
 			this.serializer = serializer;
+			this.encryptionService = encryptionService;
 			this.dataOutputStream = FileUtils.openDataOutputStream(path.toFile());
+
+			metadata = new BlueFileMetadata();
+			if (encryptionService.isEncryptionEnabled()) {
+				metadata.put(BlueFileMetadataKey.ENCRYPTION_VERSION_KEY, encryptionService.getCurrentEncryptionVersionKey());
+			}
 		} catch (IOException e) {
 			throw new BlueDbException("Failed to create BlueObjectOutput for path " + path, e);
 		}
 	}
 
 	public void writeBytes(byte[] bytes) throws BlueDbException {
```

Comment:
Good call here, another case where I was trying to avoid changing the interface when the right call was to update it to make it more readable. Glad to have the extra insight on this. I've updated the method name to be called `writeBytesAndAllowEncryption` to match the style of `writeBytesAndForceSkipEncryption`, I think those reflect that encryption isn't guaranteed to happen in either case since it is reliant on other factors. Let me know if you'd prefer a different change though, happy to discuss more!

Reply:
I did also go through  the call hierarchy and catch a few spots that seemed incorrect, good idea there.

Reply:
I like that name. How do you feel about renaming the `bytes` parameters in `writeBytesAndAllowEncryption`. Just helps make it clear that you have to pass in unencrypted bytes or you could end up with double encrypted bytes which we'd never be able to read in again.

`writeBytesAndForceSkipEncryption` is fine as is, cause you can pass in unencrypted or encrypted bytes.

Reply:
I think that is a good idea, better to be more explicit. Updated!